### ISSUE TITLE
Streaming HTTP responses

### ISSUE THREAD

**Issue by SquidDev** [SquidDev-CC/CC-Tweaked#1181]
The HTTP API currently buffers responses on the Java side, only firing a `http_success`/`http_failure` message once the response has been closed. It would be neat if there was an option to open HTTP handles in "streaming" mode, where the `http_success` event is queued immediately after headers are received. The various read methods would block until the HTTP body is received:

```lua
local response = http.get { url = "https://example.com", streaming = true }
while true do
  -- Read some data, blocking until available.
  local body = response:read(8192)
  if not body then return end

  -- Do something with this chunk of data
end
```

The main motivation for this is better support for downloading large (or possibly infinitely long) files without having to load the whole thing into memory. More complex use-cases are still better served by websockets.

Some concerns/comments:
 - I think the correct API here is that `.read()` is what's responsible for reading from the socket, yielding if there's insufficient data. This is different to the rest of CC's API; rednet and websockets passively send events - all `.receive()` is doing is listening to them.

   This is pretty similar to what [CCTweaks's tcp support](https://github.com/SquidDev-CC/CCTweaks-Lua/wiki/socket-library) did, and I wasn't wild about it then. Also haven't looked into how this would be implemented on top of Netty or, indeed, if it's possible!

 - There's a temptation here to make the Java API _only_ provide streaming sockets, and then [pre-process the events in Lua](https://github.com/cc-tweaked/CC-Tweaked/blob/5be290a1e2eabd28f4e2632034b9dba18ff1d1c6/src/main/resources/data/computercraft/lua/bios.lua#L979-L991), capturing HTTP handles before they're passed to user code.

   It'd definitely simplify implementation on the Java side, but not 100% sure. It's definitely a bit of a departure from CraftOS's current hands-off design :).

**Comment by Commandcracker** [comment]
I'm looking forward to HTTP Streaming! 😄 

> The main motivation for this is better support for downloading large (or possibly infinitely long) files without having to load the whole thing into memory. More complex use-cases are still better served by websockets.

The problem with websockets is that their max message size is `128 * 1024 = 131072` while HTTP's max download size is `16 * 1024 * 1024 = 16777216`. This means it would take me around 128 web-socket messages to download the same amount of data. This would take more time, and it is not acceptable for my use case of video streaming. It works for small screens but not if I want to stream to a max sized monitor, so I'm basically forced to use multiple HTTP requests if I want to do that. HTTP Streaming would be perfect for my use case as it removes the need of multiple requests (improving the speed).

In addition, HTTP Streaming would make it possible to have Server-Sent Events (SSE) which I'm a fan of :)
[Here](https://gist.github.com/CMCDragonkai/6bfade6431e9ffb7fe88) is a nice write-up that I found about HTTP Streaming, but it is probably more useful for using it regarding the headers.

_I might try to implement it my self, but my Java is a bit rusty. (skill issues :D)_

**Comment by zyxkad** [comment]
@SquidDev hi, recently I'm working on a git client and found this issue again.

Git smart protocol send streaming data about remote log & the actually packfile, and due to lack of streaming, my client can only display all remote logs at once after a long delay, and the entire request may fail if the git repo grow large enough.

Is there any progress / reconsideration on this? Currently I have some idea about it (like you said maintain a Lua side buffer, and yield when more data is needed), will you accept PR?

**Comment by SquidDev** [comment]
It's not something I've looked at yet, no. I think if the PR is simple, definitely happy to review it

**Comment by zyxkad** [comment]
My current attempt is queue an event `http_content` when netty channel receives data. It works well for small response body, but for larger body it will soon consume all queue space in a single tick, since netty chunk data into buffer with max 8192 bytes. Not sure what's a good solution here, my idea will be have a global connection manager that maintains a body buffer for each connection on the computers and only send one `http_content` event per tick with all data in the range combined, while we also have to have a way to limit the http download speed to the max combined buffer / tick (or just reuse the existing config option `maxDownload` but apply to per tick instead of the whole response) to prevent abuse streaming connection.

**Comment by SquidDev** [comment]
So I think the difference with HTTP streaming compared to websockets, is that the client should be in charge of when we read data. You don't want it so that all data is read from the server eagerly.  For example, if you want to stream a large file and play it to a speaker, you don't want to be continuing to download the file unbounded while speaker audio plays out[^1].


I think the design here looks something like:
 - We maintain some internal buffer of HTTP contents on the Java side.
   - If the buffer was empty, we queue some event (`http_ready`, IDK?) to wake up any readers.
   - If the buffer is full (>= `maxDownload`), then we pause reading.
 - The Lua-exposed `read()` function pulls from the buffer.
   - If the buffer is empty, we wait for our `http_ready` event (and also probably some handling of the connection being closed, etc...).
   - If the buffer was full, and no longer is, then we unpause reading over the network.

[^1]: One of the problems with websockets (and modems) is there's no model of backpressure!

**Comment by zyxkad** [comment]
> So I think the difference with HTTP streaming compared to websockets, is that the client should be in charge of when we read data.

I see your point.
Do you think we should add a native method to request for more data? i.e. we fire an `http_content` event with current buffered response (if any), and only after the method is invoked.
Or is it a better idea make the `read()` java side instead and don't fire any extra event about it? (since relying on event queue is not a good idea in large peripheral network, and they also easily broke with `sleep` or any main thread method invokes)

*right now my `http_content` event structure is close to existing `http_success/failure`: `"http_content", url, bodyData`*

---
Edit: I re-read what you said, and think you are suggesting add a `pollBody` method? and fire `http_content_ready` *(a clearer name than `http_ready`, so you don't mix it up with `http_success`)* when body buffer from empty to with some data.
Then lua impl will be something like this
```lua
local buffer = ''
function response.read(count)
  if buffer == nil then -- already EOF
    return nil
  end
  while #buffer < count do
    local data = response.pollBody()
    if not data then
      -- connection closed
      local res = buffer
      buffer = nil
      return res
    end
    buffer = buffer .. data
    if buffer >= count then
      break
    end
    os.pullEvent('http_content_ready')
  end
  local res = buffer:sub(1, count)
  buffer = buffer:sub(count + 1)
  return res
end
```

**Comment by SquidDev** [comment]
No. I don't think it makes sense to implement this any of this in Lua — then you'd have to intercept the `http_success`/`http_failure` events and inject in the additional methods. We don't need a separate `pollBody` method — this can just be done inside `read`.

**Comment by zyxkad** [comment]
@SquidDev I noticed AbstractHandle is allocating direct buffer here, I wonder if it is a misuse? since the channel we are reading (at least for http request, not sure about file system) is using java heap `byte[]` as backing, not direct network buffer.

https://github.com/cc-tweaked/CC-Tweaked/blob/663ffed4337da0dc3d82ace1e813e3c78b4a8c99/projects/core/src/main/java/dan200/computercraft/core/apis/handles/AbstractHandle.java#L149
